tests/lax_control_flow_test.py

```diff
@@ -2067,6 +2067,62 @@ def linear_solve(a, b):
     actual = api.vmap(linear_solve, (None, 1), 1)(a, c)
     self.assertAllClose(expected, actual)
 
+  @jtu.skip_on_flag("jax_skip_slow_tests", True)
+  def test_custom_linear_solve_aux(self):
+    def explicit_jacobian_solve_aux(matvec, b):
+      x = lax.stop_gradient(jnp.linalg.solve(api.jacobian(matvec)(b), b))
+      return x, orig_aux
+
+    def matrix_free_solve_aux(matvec, b):
+      return lax.custom_linear_solve(
+        matvec, b, explicit_jacobian_solve_aux, explicit_jacobian_solve_aux,
+        symmetric=True, has_aux=True)
+
+    def linear_solve_aux(a, b):
+      return matrix_free_solve_aux(partial(lax.dot, a), b)
+
+    orig_aux = {"converged": True, "nfev": 12345}
+    rng = np.random.RandomState(0)
+    a = rng.randn(3, 3)
+    a = a + a.T
+    b = rng.randn(3)
+
+    expected = jnp.linalg.solve(a, b)
+    expected_jit_aux = tree_util.tree_map(np.asarray, orig_aux)
+    actual_nojit, nojit_aux = linear_solve_aux(a, b)
+    actual_jit, jit_aux = api.jit(linear_solve_aux)(a, b)
+
+    self.assertAllClose(expected, actual_nojit)
+    self.assertAllClose(expected, actual_jit)
+    # scalar dict equality check
+    self.assertDictEqual(nojit_aux, orig_aux)
+    self.assertDictEqual(jit_aux, expected_jit_aux)
+
+    # jvp test by hand
+    primals = (a, b)
+    solve_jvp = partial(api.jvp, linear_solve_aux)
+    tangents = tree_util.tree_map(partial(jtu.rand_like, rng), primals)
+    v_out, t_out = solve_jvp(primals, tangents)
+    v_val, v_aux = v_out
+    t_val, t_aux = t_out
+
+    self.assertAllClose(v_val, expected)
+    self.assertDictEqual(v_aux, orig_aux)
+
+    # verify symbolical zero gradient
+    t_aux_flat, _ = tree_util.tree_flatten(t_aux)
+    self.assertListEqual(t_aux_flat, [np.zeros((), dtype=jax.dtypes.float0)] * len(orig_aux))
+
+    # vmap test
+    c = rng.randn(3, 2)
+    expected = jnp.linalg.solve(a, c)
+    expected_aux = tree_util.tree_map(partial(np.repeat, repeats=2), orig_aux)
+    actual_vmap, vmap_aux = api.vmap(linear_solve_aux, (None, 1), -1)(a, c)
+
+    self.assertAllClose(expected, actual_vmap)
+    jtu.check_eq(expected_aux, vmap_aux)
+
```

Comment:
Maybe add `grad` test, too, just to make sure things don't break?

Reply:
I now opted to use array-like aux arguments, which enables the use of `jtu.check_grads` - I follow the same procedure now as in the test directly above.